src/buckets.ts

```diff
@@ -0,0 +1,253 @@
+      // Only process the first  input if there are more than one
+      const source: File | undefined = (await normaliseInput(input).next()).value
+      const client = grpc.client<pb.PushPathRequest, pb.PushPathReply, APIPushPath>(API.PushPath, {
+        host: this.serviceHost,
+        transport: this.rpcOptions.transport,
+        debug: this.rpcOptions.debug,
+      })
+      client.onMessage((message) => {
+        if (message.hasError()) {
+          // Reject on first error
+          reject(new Error(message.getError()))
+        } else if (message.hasEvent()) {
+          const event = message.getEvent()?.toObject()
+          if (event?.path) {
+            // @todo: Is there an standard library/tool for this step in JS?
+            const pth = event.path.startsWith('/ipfs/') ? event.path.split('/ipfs/')[1] : event.path
+            const cid = new CID(pth)
+            const res: PushPathResult = {
+              path: {
+                path: `/ipfs/${cid.toString()}`,
+                cid: cid,
+                root: cid,
+                remainder: '',
+              },
+              root: event.root?.path ?? '',
+            }
+            resolve(res)
+          } else if (opts?.progress) {
+            opts.progress(event?.bytes)
+          }
+        } else {
+          reject(new Error('Invalid reply'))
+        }
+      })
+      client.onEnd((code) => {
+        if (code === grpc.Code.OK) {
+          resolve()
+        } else {
+          reject(new Error(code.toString()))
+        }
+      })
+      if (source) {
+        const head = new pb.PushPathRequest.Header()
+        head.setPath(source.path || path)
+        head.setKey(key)
+        const req = new pb.PushPathRequest()
+        req.setHeader(head)
+        client.start(this.context.withContext(ctx).toJSON())
+        client.send(req)
+
+        if (source.content) {
+          for await (const chunk of source.content) {
+            const part = new pb.PushPathRequest()
+            part.setChunk(chunk as Buffer)
+            client.send(part)
+          }
+          client.finishSend()
+        }
+      }
+    })
+  }
+
+  /**
+   * Pulls the bucket path, returning the bytes of the given file.
+   * @param key Unique (IPNS compatible) identifier key for a bucket.
+   * @param path A file/object (sub)-path within a bucket.
+   * @param ctx Context object containing web-gRPC headers and settings.
+   * @param opts Options to control response stream. Currently only supports a progress function.
+   */
+  pullPath(
+    key: string,
+    path: string,
+    ctx?: Context,
+    opts?: { progress?: (num?: number) => void },
+  ): AsyncIterableIterator<Uint8Array> {
+    const metadata = this.context.withContext(ctx).toJSON()
+    const chan = new Channel<Uint8Array>()
+    const request = new pb.PullPathRequest()
+    request.setKey(key)
+    request.setPath(path)
+    let written = 0
+    const response = grpc.invoke(API.PullPath, {
```

Comment:
This is shorthand for creating a client and calling the pullPath method. We _could_ also do it that way, but this felt cleaner.